### PR TITLE
Implement leveled logging

### LINKED ISSUE #254: Verbosity controls for MetalLB logs
MetalLB's logs right now are excellent for me, the developer, because they spew a _ton_ of information about what is going on. However, that's too much for some users, and they want less verbosity.

The log library we use has support for structured logging and filtering based on log level. We could use that.

### LINKED ISSUE #655: Any way to reduce log level/verbosity
Hi there,

currently we're using metallb with 5 k8s clusters, ~30 nodes/metallb-speakers overall. Each speaker writes every 5-10 minutes two log recrods for every service in cluster. For example:
```
{"caller":"main.go:267","event":"startUpdate","msg":"start of service update","service":"foo-namespace/bar-service","ts":"2020-07-01T15:22:12.278029276Z"}
{"caller":"main.go:271","event":"endUpdate","msg":"end of service update","service":"foo-namespace/bar-service","ts":"2020-07-01T15:22:12.278091447Z"}
{"caller":"main.go:267","event":"startUpdate","msg":"start of service update","service":"foo-namespace/bar-service","ts":"2020-07-01T15:29:14.991299133Z"}
{"caller":"main.go:271","event":"endUpdate","msg":"end of service update","service":"foo-namespace/bar-service","ts":"2020-07-01T15:29:14.991318263Z"}
```
So we have about 1.8 million such log records every day.

Is there any config option to reduce log level to filter out such log records? Tried to dig into source code with no success.
Maybe we have some misconfig and that situation with logs is abnormal? Config from one of clusters:
```
peers:
- peer-address: x.x.x.x
  peer-asn: xxx
  my-asn: xxx
  hold-time: 15s
  node-selectors:
  - match-labels:
      metallb: "true"
address-pools:
- name: pool-foobar
  protocol: bgp
  avoid-buggy-ips: true
  addresses:
  - x.x.x.0/24
  bgp-advertisements:
  - aggregation-length: 24
```

Kubernetes v1.16.7
kube-proxy in iptables mode
Flannel v0.11.0 in host-gw mode w/o cni
MetalLB v0.9.3 BGP/L2 depending on cluster.

### PR DESCRIPTION
**Pull request by utkuozdemir**
Signed-off-by: Utku Ozdemir <uoz@protonmail.com>

This PR implements leveled logging. I set the levels as I've seen fit by looking at their messages - any feedbacks are appreciated.
Fixes #254 
Fixes #655

### CODE REVIEW

**Comment by gclawes** [comment]
Should we update the helm charts and manifests to add the `--log-level` flag?

**Comment by utkuozdemir** [comment]
> Should we update the helm charts and manifests to add the `--log-level` flag?

Agreed and added.

What do you think about using `info` as the default value? On one hand it's a reasonable default, on the other hand if you want to keep the logging behavior unchanged, `all` or `debug` can do that.

**Comment by champtar** [comment]
On my phone, but I think I would bump many debug to info, all the ones that show a change in the system. The most spam comes from memberlist, then service sync I think.

**Comment by utkuozdemir** [comment]
@champtar 
Thanks. Please see my latest commit which bumps some of the logs from debug to info.
With this change, debug level is used for logs that tell about:
- the starts and ends of operations
- no-ops ("nothing to do"s)

The operations that inform about a change/update etc. are now using the info level.

File: internal/speakerlist/speakerlist.go [hunk 1]
```diff
@@ -74,7 +75,7 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "Memberlist"))
 	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
```
Comment (champtar) [inline internal/speakerlist/speakerlist.go:76]: Those 2 lines generate logs like:
```
{"caller":"net.go:210","component":"Memberlist","msg":"[DEBUG] memberlist: Stream connection from=10.10.52.152:51718","ts":"2021-07-07T20:33:58.686178124Z"}
{"caller":"net.go:785","component":"Memberlist","msg":"[DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.153:7946","ts":"2021-07-07T20:34:26.625477039Z"}
{"caller":"net.go:210","component":"Memberlist","msg":"[DEBUG] memberlist: Stream connection from=10.10.52.152:52002","ts":"2021-07-07T20:34:28.689981448Z"}
{"caller":"net.go:210","component":"Memberlist","msg":"[DEBUG] memberlist: Stream connection from=10.10.52.153:38596","ts":"2021-07-07T20:34:45.650096602Z"}
{"caller":"net.go:785","component":"Memberlist","msg":"[DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.153:7946","ts":"2021-07-07T20:34:56.628673276Z"}
```

Is it possible from `[DEBUG]` to set the level to debug ?

Comment (champtar) [reply internal/speakerlist/speakerlist.go:76]: https://github.com/hashicorp/memberlist/search?q=logger.Printf

Comment (utkuozdemir) [reply internal/speakerlist/speakerlist.go]: It was possible with some c/p from the `NewStdlibAdapter` and adapting it to extract the level from the logs submitted by memberlist. Please have a look to the last commit.

Comment (champtar) [reply internal/speakerlist/speakerlist.go]: Thanks !

File: controller/main.go [hunk 2]
```diff
@@ -17,9 +17,11 @@ package main
 import (
 	"flag"
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline controller/main.go:20]: move it with github.com/go-kit/kit/log

File: controller/service.go [hunk 3]
```diff
@@ -16,6 +16,7 @@ package main
 
 import (
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline controller/service.go:19]: move it with github.com/go-kit/kit/log

File: controller/service.go [hunk 4]
```diff
@@ -41,7 +42,7 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 	// ipFamily to use.
 	clusterIP := net.ParseIP(svc.Spec.ClusterIP)
 	if clusterIP == nil {
-		l.Log("event", "clearAssignment", "reason", "noClusterIP", "msg", "No ClusterIP")
+		level.Debug(l).Log("event", "clearAssignment", "reason", "noClusterIP", "msg", "No ClusterIP")
```
Comment (champtar) [inline controller/service.go:45]: Info() ?

File: internal/bgp/bgp.go [hunk 5]
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/bgp/bgp.go:9]: move it with github.com/go-kit/kit/log

File: internal/k8s/k8s.go [hunk 6]
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/base64"
 	"errors"
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/k8s/k8s.go:9]: move it with github.com/go-kit/kit/log

File: internal/layer2/announcer.go [hunk 7]
```diff
@@ -1,6 +1,7 @@
 package layer2
 
 import (
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/layer2/announcer.go:4]: move it with github.com/go-kit/kit/log

File: internal/layer2/arp.go [hunk 8]
```diff
@@ -3,6 +3,7 @@ package layer2
 import (
 	"bytes"
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/layer2/arp.go:6]: move it with github.com/go-kit/kit/log

File: internal/layer2/arp.go [hunk 9]
```diff
@@ -99,10 +100,10 @@ func (a *arpResponder) processRequest() dropReason {
 	}
 
 	stats.GotRequest(pkt.TargetIP.String())
-	a.logger.Log("interface", a.intf, "ip", pkt.TargetIP, "senderIP", pkt.SenderIP, "senderMAC", pkt.SenderHardwareAddr, "responseMAC", a.hardwareAddr, "msg", "got ARP request for service IP, sending response")
+	level.Info(a.logger).Log("interface", a.intf, "ip", pkt.TargetIP, "senderIP", pkt.SenderIP, "senderMAC", pkt.SenderHardwareAddr, "responseMAC", a.hardwareAddr, "msg", "got ARP request for service IP, sending response")
```
Comment (champtar) [inline internal/layer2/arp.go:103]: Debug() ?

File: internal/layer2/ndp.go [hunk 10]
```diff
@@ -2,6 +2,7 @@ package layer2
 
 import (
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/layer2/ndp.go:5]: move it with github.com/go-kit/kit/log

File: internal/layer2/ndp.go [hunk 11]
```diff
@@ -137,10 +138,10 @@ func (n *ndpResponder) processRequest() dropReason {
 	}
 
 	stats.GotRequest(ns.TargetAddress.String())
-	n.logger.Log("interface", n.intf, "ip", ns.TargetAddress, "senderIP", src, "senderLLAddr", nsLLAddr, "responseMAC", n.hardwareAddr, "msg", "got NDP request for service IP, sending response")
+	level.Info(n.logger).Log("interface", n.intf, "ip", ns.TargetAddress, "senderIP", src, "senderLLAddr", nsLLAddr, "responseMAC", n.hardwareAddr, "msg", "got NDP request for service IP, sending response")
```
Comment (champtar) [inline internal/layer2/ndp.go:141]: Debug() ?

File: internal/logging/logging.go [hunk 12]
```diff
@@ -6,6 +6,7 @@ import (
 	"bufio"
 	"flag"
 	"fmt"
+	loglevel "github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/logging/logging.go:9]: move it with github.com/go-kit/kit/log

File: internal/speakerlist/speakerlist.go [hunk 13]
```diff
@@ -14,7 +14,7 @@ package speakerlist
 
 import (
 	"crypto/sha256"
-	golog "log"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline internal/speakerlist/speakerlist.go:17]: move it with github.com/go-kit/kit/log

Comment (champtar) [reply internal/speakerlist/speakerlist.go:17]: Also `gokitlog "github.com/go-kit/kit/log"` -> `"github.com/go-kit/kit/log"`

File: speaker/bgp_controller.go [hunk 14]
```diff
@@ -16,6 +16,7 @@ package main
 
 import (
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline speaker/bgp_controller.go:19]: move it with github.com/go-kit/kit/log

File: speaker/main.go [hunk 15]
```diff
@@ -17,10 +17,12 @@ package main
 import (
 	"flag"
 	"fmt"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline speaker/main.go:20]: move it with github.com/go-kit/kit/log

**Review by champtar — CHANGES_REQUESTED** [review]
Some nitpicks
After those changes LGTM

File: controller/main.go (hunk 2 above)
Comment (utkuozdemir) [reply controller/main.go:20]: > Some nitpicks
> After those changes LGTM

sure, will take care of those small changes soon. Just one thing: I didn't get what you mean in your comments by
> move it with github.com/go-kit/kit/log

Can you explain?
Thanks.

Comment (champtar) [reply controller/main.go:20]: Put the 2 gokit imports one after the other

**Comment by utkuozdemir** [comment]
@champtar applied the changes you requested in the last 2 commits. Please check.

File: controller/service.go [hunk 16]
```diff
@@ -32,7 +32,7 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 	// Not a LoadBalancer, early exit. It might have been a balancer
 	// in the past, so we still need to clear LB state.
 	if svc.Spec.Type != "LoadBalancer" {
-		level.Debug(l).Log("event", "clearAssignment", "reason", "notLoadBalancer", "msg", "not a LoadBalancer")
+		level.Info(l).Log("event", "clearAssignment", "reason", "notLoadBalancer", "msg", "not a LoadBalancer")
```
Comment (champtar) [inline controller/service.go:35]: This one should stay debug else we list all services

File: controller/main.go [hunk 17]
```diff
@@ -29,8 +28,10 @@ import (
 	"go.universe.tf/metallb/internal/logging"
 	"go.universe.tf/metallb/internal/version"
 
-	"github.com/go-kit/kit/log"
 	v1 "k8s.io/api/core/v1"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
```
Comment (champtar) [inline controller/main.go:34]: More nitpick, but I would not move `"github.com/go-kit/kit/log"`, just put `"github.com/go-kit/kit/log/level"` after it.
the less diff you have the easier it is for people to rebase on top

**Review by champtar — CHANGES_REQUESTED** [review]

**Comment by champtar** [comment]
And if you can squash your commits a bit that would be perfect :)

**Review by champtar — APPROVED** [review]
not tested but LGTM !

**Review by utkuozdemir — COMMENTED** [review]
_

File: charts/metallb/values.yaml [hunk 18]
```diff
@@ -127,6 +127,8 @@ prometheus:
 # controller.
 controller:
   enabled: true
+  # -- Controller log level. Must be one of: `all`, `debug`, `info`, `warn`, `error` or `none`
+  logLevel: info
```
Comment (russellb) [inline charts/metallb/values.yaml:131]: How about changing the manifests in `manifests/` to match the helm chart default value of `info` ?

File: controller/main.go [hunk 19]
```diff
@@ -81,32 +83,32 @@ func (c *controller) SetBalancer(l log.Logger, name string, svcRo *v1.Service, _
 		st, svc = svc.Status, svcRo.DeepCopy()
 		svc.Status = st
 		if err := c.client.UpdateStatus(svc); err != nil {
-			l.Log("op", "updateServiceStatus", "error", err, "msg", "failed to update service status")
+			level.Error(l).Log("op", "updateServiceStatus", "error", err, "msg", "failed to update service status")
 			return k8s.SyncStateError
 		}
 	}
-	l.Log("event", "serviceUpdated", "msg", "updated service object")
+	level.Info(l).Log("event", "serviceUpdated", "msg", "updated service object")
```
Comment (russellb) [inline controller/main.go:90]: Unless we were to expand this to give more info (what service was updated? how was it updated?), i'd probably keep this as debug to match the other messages it's related to.

Comment (russellb) [reply controller/main.go:90]: Maybe we should file an issue to go back through log messages to improve them to include which resource is being looked at / acted upon?

File: controller/main.go [hunk 20]
```diff
@@ -81,32 +83,32 @@ func (c *controller) SetBalancer(l log.Logger, name string, svcRo *v1.Service, _
...
 
 	return k8s.SyncStateSuccess
 }
 
 func (c *controller) deleteBalancer(l log.Logger, name string) {
 	if c.ips.Unassign(name) {
-		l.Log("event", "serviceDeleted", "msg", "service deleted")
+		level.Info(l).Log("event", "serviceDeleted", "msg", "service deleted")
```
Comment (russellb) [inline controller/main.go:97]: This seems like a debug message to me.

Unrelated to your change, this message also could be misleading. MetalLB didn't delete the service. We're just emitting a debug message saying that we saw a service get deleted. The text could be improved ...

**Comment by champtar** [comment]
@russellb did you delete some of your reviews ? I got email notification that I can't in here

**Comment by russellb** [comment]
> @russellb did you delete some of your reviews ? I got email notification that I can't find in here

yeah, sorry.  Commented on some, changed my mind, deleted, and now going back through and will post them all together as a single "review".  I probably should have just replied to myself on here.

File: charts/metallb/values.yaml (hunk 18 above)
Comment (utkuozdemir) [reply charts/metallb/values.yaml:131]: Makes sense, added.

File: internal/k8s/k8s.go [hunk 21]
```diff
@@ -189,12 +190,12 @@ func New(cfg *Config) (*Client, error) {
 
 				c.syncFuncs = append(c.syncFuncs, c.epInformer.HasSynced)
 			} else {
-				c.logger.Log("op", "New", "msg", "using endpoint slices")
+				level.Info(c.logger).Log("op", "New", "msg", "using endpoint slices")
```
Comment (russellb) [inline internal/k8s/k8s.go:193]: I'd go with debug here

**Review by russellb — COMMENTED** [review]
This looks good to me.  I spotted one additional message I would have chosen debug, but the rest looked good.  I started taking notes on messages I'd like to change in follow-ups but stopped since none of them are really related to these changes.

File: internal/k8s/k8s.go (hunk 21 above)
Comment (champtar) [reply internal/k8s/k8s.go:193]: Endpoint slices are still new in MetalLB so it's a good information to have in bug report for now IMO.

Comment (champtar) [reply internal/k8s/k8s.go:193]: (and it's just 1 line)

Comment (russellb) [reply internal/k8s/k8s.go:193]: I guess so, but it's an implementation detail that shouldn't matter unless something is broken (like many other debug logs).

Like you said, it's just 1 line so I don't feel that strongly ...

**Comment by russellb** [comment]
... CI doesn't seem to have run on this PR?  I have no idea why.

**Comment by russellb** [comment]
I pushed a rebase of this PR to the branch to try to trigger CI and it seems to have worked

File: internal/k8s/k8s.go (hunk 21 above)
Comment (utkuozdemir) [reply internal/k8s/k8s.go:192]: Should I do this change? Or you can do it directly.

**Comment by utkuozdemir** [comment]
Some of my changes seems to have disappeared - I think it was caused by the rebase somehow, but not sure. I have (re)pushed those changes in the latest commit.

**Comment by russellb** [comment]
> Some of my changes seems to have disappeared - I think it was caused by the rebase somehow, but not sure. I have (re)pushed those changes in the latest commit.

I apologize if I overwrote something. I thought I was up to date from your branch before I did the rebase.  In any case, it seems CI is running (and passing) now.  FWIW, I saw those same failures locally when I pulled down this PR to test it, so maybe those build fixes hadn't been pushed?

**Comment by utkuozdemir** [comment]
> > Some of my changes seems to have disappeared - I think it was caused by the rebase somehow, but not sure. I have (re)pushed those changes in the latest commit.
> 
> I apologize if I overwrote something. I thought I was up to date from your branch before I did the rebase. In any case, it seems CI is running (and passing) now. FWIW, I saw those same failures locally when I pulled down this PR to test it, so maybe those build fixes hadn't been pushed?

No problem at all.

I just think `internal/speakerlist/log.go` went missing, because I added it as @champtar 's request, and I think he saw it: https://github.com/metallb/metallb/pull/895#discussion_r665764127 

Or it might be my squash that made it disappear 🤔 

But anyway, doesn't matter - all back in their place now :)

**Comment by russellb** [comment]
I'm trying to build this locally to look at logs and it's failing.  Any idea why?

```
$ inv dev-env -p layer2
kind version
kind v0.10.0 go1.15.7 linux/amd64
go build -v -o build/amd64/controller/controller -ldflags '-X go.universe.tf/metallb/internal/version.gitCommit=25e1592a -X go.universe.tf/metallb/internal/version.gitBranch=pr-895' go.universe.tf/metallb/controller
go.universe.tf/metallb/internal/logging
# go.universe.tf/metallb/internal/logging
internal/logging/logging.go:56:14: level.NewFilter undefined (type string has no field or method NewFilter)
internal/logging/logging.go:147:15: level.AllowAll undefined (type string has no field or method AllowAll)
internal/logging/logging.go:149:15: level.AllowDebug undefined (type string has no field or method AllowDebug)
internal/logging/logging.go:151:15: level.AllowInfo undefined (type string has no field or method AllowInfo)
internal/logging/logging.go:153:15: level.AllowWarn undefined (type string has no field or method AllowWarn)
internal/logging/logging.go:155:15: level.AllowError undefined (type string has no field or method AllowError)
internal/logging/logging.go:157:15: level.AllowNone undefined (type string has no field or method AllowNone)
```

```
$ go version
go version go1.16.4 linux/amd64
```

**Comment by utkuozdemir** [comment]
It was fixed by my last commit: https://github.com/metallb/metallb/commit/93208e24614fba55348f9a7cbd8fd9bf2077ed97

Specifically, by the rename `level` -> `lvl`.

Can you delete your local branch and checkout again the remote branch? Might help.

**Comment by russellb** [comment]
> It was fixed by my last commit: [93208e2](https://github.com/metallb/metallb/commit/93208e24614fba55348f9a7cbd8fd9bf2077ed97)
> 
> Specifically, by the rename `level` -> `lvl`.
> 
> Can you delete your local branch and checkout again the remote branch? Might help.

oops, you're right.  :-D  thanks

**Comment by russellb** [comment]
OK, last request, I think ... would you mind adding an entry to the "New Features" section of the release notes about this?  I would note that leveled logging is now supported, what the default behavior using our helm chart or manifests, and how someone would go about changing it as needed.

See `website/content/release-notes/_index.md`

**Comment by utkuozdemir** [comment]
> OK, last request, I think ... would you mind adding an entry to the "New Features" section of the release notes about this? I would note that leveled logging is now supported, what the default behavior using our helm chart or manifests, and how someone would go about changing it as needed.
> 
> See `website/content/release-notes/_index.md`

Done, but feel free to change the wording as you see fit.

**Comment by russellb** [comment]
> > OK, last request, I think ... would you mind adding an entry to the "New Features" section of the release notes about this? I would note that leveled logging is now supported, what the default behavior using our helm chart or manifests, and how someone would go about changing it as needed.
> > See `website/content/release-notes/_index.md`
> 
> Done, but feel free to change the wording as you see fit.

looks good to me.  Thank you for all of your work!

**Comment by utkuozdemir** [comment]
Thanks! Happy to contribute :)

**Comment by zigmund** [comment]
@utkuozdemir, thanks for the fix.

@russellb, are you going to release this in near future?

**Comment by russellb** [comment]
I'll probably wait a bit.  We just did a feature release (0.10) pretty recently, so I figure we'd wait a bit longer to get some other stuff in first.

**Comment by cjreyn** [comment]
Thank you for all the good work on Metallb. We're using it heavily to support science in the UK at Diamond Light Source!
Do you have any idea which release this PR is planned for?

**Comment by curantes** [comment]
> Thank you for all the good work on Metallb. We're using it heavily to support science in the UK at Diamond Light Source!
> Do you have any idea which release this PR is planned for?

v0.11.0 according to the milestone and release date have not been announced what I know of. 

disclaimer: I am not part of metallb project

**Comment by zigmund** [comment]
It's really long-waited fix for big amounts of users, please release.

**Pull request merged**